pgjdbc/src/test/java/org/postgresql/test/jdbc2/SQLDataTest.java

```diff
@@ -0,0 +1,396 @@
+
+    TestUtil.createTable(con, TABLE_TEST, String.format("%s int, thingid int", COL_ID));
+    stmt.executeUpdate(String.format("INSERT INTO %s VALUES (1, 42)", TABLE_TEST));
+
+    TestUtil.createCompositeType(con, UDT_THING, String.format("%s int, thing %s", COL_ID, TABLE_THING));
+    TestUtil.createCompositeType(con, UDT_THING_COL, String.format("name text, things %s[]", TABLE_THING));
+
+    TestUtil.closeQuietly(stmt);
+    TestUtil.closeDB(con);
+  }
+
+  @AfterAll
+  public static void tearDown() throws Exception {
+    Connection con = TestUtil.openDB();
+    TestUtil.dropTable(con, TABLE_THING);
+    TestUtil.dropTable(con, TABLE_TEST);
+    TestUtil.dropType(con, UDT_THING);
+    TestUtil.dropType(con, UDT_THING_COL);
+    TestUtil.closeDB(con);
+  }
+
+  @BeforeEach
+  private void beforeEach() throws Exception {
+    con = TestUtil.openDB();
+    stmt = con.createStatement();
+  }
+
+  @AfterEach
+  private void afterEach() throws Exception {
+    TestUtil.closeQuietly(stmt);
+    TestUtil.closeDB(con);
+  }
+
+  private ResultSet getResultSet(String sql) throws SQLException {
+    ResultSet rs = stmt.executeQuery(sql);
+    assertTrue(rs.next());
+    return rs;
+  }
+
+  private void checkThing(Thing thing) {
+    assertEquals(42, thing.id);
+    assertEquals(43, thing.idl);
+    assertEquals(44, thing.ids);
+    assertEquals(NAME, thing.name);
+    assertTrue(thing.valid);
+    assertEquals(1, thing.bytey);
+    assertEquals(FLOATY, thing.floaty);
+    assertEquals(DOUBLY, thing.doubly);
+    assertEquals(BIGD, thing.bigD);
+    assertArrayEquals(BYTES.getBytes(), thing.bytes);
+    assertEquals(DATE, thing.date.toString());
+    assertEquals(TIME, thing.time.toString());
+    assertEquals(TS, thing.timestamp.toString());
+  }
+
+  private void checkNullThing(Thing thing) {
+    assertEquals(0, thing.id);
+    assertEquals(0, thing.idl);
+    assertEquals(0, thing.ids);
+    assertNull(thing.name);
+    assertFalse(thing.valid);
+    assertEquals(0, thing.bytey);
+    assertEquals(0, thing.floaty);
+    assertEquals(0, thing.doubly);
+    assertNull(thing.bigD);
+    assertNull(thing.bytes);
+    assertNull(thing.date);
+    assertNull(thing.time);
+    assertNull(thing.timestamp);
+  }
+
+  static Thing[] toThingArray(Array array) throws SQLException {
+    Map<String, Class<?>> map = new HashMap<>();
+    map.put(TABLE_THING, Thing.class);
+
+    Object[] objects = (Object[])array.getArray(map);
+
+    Thing[] things = new Thing[objects.length];
+    for (int ii = 0; ii < objects.length; ii++) {
+      things[ii] = (Thing) objects[ii];
+    }
+    return things;
+  }
+
+  private void checkThings(Thing[] things) {
+    assertNotNull(things);
+    assertEquals(2, things.length);
+
+    for (Thing thing : things) {
+      if (thing.id == 0) {
+        checkNullThing(thing);
+      } else {
+        checkThing(thing);
+      }
+    }
+  }
+
+  // private void parseObj(String value) {
```

Comment:
is there a reason to keep this here?

Reply:
No, was just keeping until this was all resolved. Cheers.